File: Classes/Body Parts/Ears.vb
Imports CoreAudioApi

''' <summary>
''' The Ears class allows us to hear fish by monitoring the
''' output from the sound card.  
''' </summary>
Public Class Ears

    Public Event HearsAFish()

    Private WithEvents SndDevice As MMDevice
    Private WithEvents tmeTimer As System.Windows.Forms.Timer
    Private VolumeQueue As New Queue(Of Integer)

    ' ToDo - Allow the user to set these values through the GUI
    Private Const MAX_VOLUME_QUEUE_LENGTH As Integer = 5
    Private Const SOUND_TOLERANCE_LEVEL As Integer = 15

    ''' <summary>
    ''' Create a new 'set of ears' for listening
    ''' </summary>    
    Public Sub New()

        ' Setup our sound listener
        Dim SndDevEnum As New MMDeviceEnumerator()
        SndDevice = SndDevEnum.GetDefaultAudioEndpoint(EDataFlow.eRender, ERole.eMultimedia)

        ' Setup our timer
        tmeTimer = New System.Windows.Forms.Timer
        tmeTimer.Interval = 500
        AddHandler tmeTimer.Tick, AddressOf tmeTimer_Tick
        tmeTimer.Enabled = True

    End Sub

    ''' <summary>
    ''' Get the current sound output level
    ''' </summary>    
    Private Sub tmeTimer_Tick()

        ' Get the current level
        Dim currentVolumnLevel As Integer = CType(SndDevice.AudioMeterInformation.MasterPeakValue * 100, Integer)
        VolumeQueue.Enqueue(currentVolumnLevel)

        ' Keep a running queue of the last X sounds as a reference point
        If VolumeQueue.Count >= MAX_VOLUME_QUEUE_LENGTH Then
            VolumeQueue.Dequeue()
        End If

        ' Determine if the current level is high enough to be a fish
        If currentVolumnLevel - GetAverageVolume() >= My.Settings.SplashLimit Then
            RaiseEvent HearsAFish()
        End If
    End Sub

    ''' <summary>
    ''' Returns the average volumn level in our sound queue
    ''' </summary>
    Private Function GetAverageVolume() As Integer
        Return CType(VolumeQueue.Sum / VolumeQueue.Count, Integer)
    End Function
End Class

File: UltimateFishBot/Classes/BodyParts/Ears.cs
using CoreAudioApi;
using System;
using System.Diagnostics;
using System.Linq;
using System.Threading;
using System.Threading.Tasks;
using UltimateFishBot.Collections;

namespace UltimateFishBot.BodyParts
{
    internal class Ears
    {
        private MMDevice _sndDevice;
        private readonly LimitedCollection<int> _volumeQueue;
        private int _tickrate = 50; //ms pause between sound checks

        private const int MaxVolumeQueueLength = 5;

        public Ears()
        {
            _volumeQueue = new LimitedCollection<int>(MaxVolumeQueueLength) {0};
        }

        public async Task<bool> Listen(int millisecondsToListen, CancellationToken cancellationToken)
        {
            var stopwatch = new Stopwatch();
            stopwatch.Start();
            var sndDevEnum = new MMDeviceEnumerator();
            _sndDevice = Properties.Settings.Default.AudioDevice != ""
                ? sndDevEnum.GetDevice(Properties.Settings.Default.AudioDevice)
                : sndDevEnum.GetDefaultAudioEndpoint(EDataFlow.eRender, ERole.eMultimedia);
            Func<bool> heardFish;
            if (Properties.Settings.Default.AverageSound)
                heardFish = ListenTimerTickAvg;
            else
                heardFish = ListenTimerTick;

            while (stopwatch.ElapsedMilliseconds <= millisecondsToListen) {
                await Task.Delay(_tickrate, cancellationToken);
                if (heardFish()) {
                    return true;
                }
            }
            return false;
        }

        private bool ListenTimerTick() {
            // Get the current level
            var currentVolumnLevel = (int)(_sndDevice.AudioMeterInformation.MasterPeakValue * 100);

            if (currentVolumnLevel >= Properties.Settings.Default.SplashLimit)
                return true;

            return false;
        }

        private bool ListenTimerTickAvg()
        {
            // Get the current level
            var currentVolumnLevel = (int)(_sndDevice.AudioMeterInformation.MasterPeakValue * 100);
            var avgVol = GetAverageVolume();
            var hear = false;

            // Determine if the current level is high enough to be a fish
            if (currentVolumnLevel - avgVol >= Properties.Settings.Default.SplashLimit) {
                Serilog.Log.Information("Hear: {av},{cvl},{queue}", avgVol, currentVolumnLevel, _volumeQueue);
                hear = true;
            }

            // Keep a running queue of the last X sounds as a reference point
            _volumeQueue.Add(currentVolumnLevel);
            return hear;

        }

        private int GetAverageVolume()
        {
            return (int)_volumeQueue.Average();
        }
    }
}
